README.md
<p align="center"><img src="https://raw.githubusercontent.com/bitcoinphantom/bitcoinphantom/master/readme-header.png" alt="bitcoinphantom Source Code"></p>
<p align="center">
  <a href="https://circleci.com/gh/bitcoinphantom/bitcoinphantom/tree/master" target="_blank"><img src="https://circleci.com/gh/bitcoinphantom/bitcoinphantom/tree/master.svg?style=svg"></a>
  <img src="https://img.shields.io/badge/status-stable-green.svg">
  <img src="https://img.shields.io/badge/license-MIT-blue.svg">
  <a href="https://codecov.io/gh/bitcoinphantom/bitcoinphantom">
    <img src="https://codecov.io/gh/bitcoinphantom/bitcoinphantom/branch/develop/graph/badge.svg" />
  </a>
  <a href="https://github.com/bitcoinphantom/bitcoinphantom/releases">
    <img alt="GitHub All Releases" src="https://img.shields.io/github/downloads/bitcoinphantom/bitcoinphantom/total?style=social">
  </a>
  <img alt="GitHub commit activity" src="https://img.shields.io/github/commit-activity/m/bitcoinphantom/bitcoinphantom">
</p>

# bitcoinphantom Source Code [btcx]

## Specifications
Specification | Value
--- | ---
Protocol | PoW (proof of Work)
Algorithms | scrypt, x17, Lyra2rev2, myr-groestl, & blake2s
Blocktime | 0.1 seconds
Total Supply | 100,000,000,000 btcx
RPC port | 11207
P2P port | 11211
pre-mine | N/A
ICO | N/A

## Blockrewards
Block Number Range | Reward
--- | ---
0 to 14,000 | 600,000 coins
14,001 to 28,000 | 300,000 coins
28,001 to 42,000 | 200,000 coins
42,001 to 210,000 | 100,000 coins
210,001 to 378,000 | 60,000 coins
378,001 to 546,000 | 37,500 coins
546,001 to 714,000 | 18,750 coins
714,001 to 2,124,000 | 4,375 coins
2,124,001 to 3,700,000 | 3868 coins
3,700,001 to 4,200,000 | 2434 coins
4,200,001 to 4,700,000 | 1217 coins
4,700,001 to 5,200,000 | 609 coins
5,200,001 to 5,700,000 | 305  coins
5,700,001 to 6,200,000 | 153  coins
6,200,001 to 6,700,000 | 77 coins
6,700,001 to 7,200,000 | 39 coins

## Resources

* [Blockchain Explorer](https://bitcoinphantom-blockchain.info/)
* [Mining Pool List](https://bitcoinphantom.com/community/btcx-mining-pools/)
* [Black Paper](https://bitcoinphantom.com/static/blackpaper/bitcoinphantom-Anonymity-Centric-CryptoCurrency.pdf)

### Community

* [Telegram](https://t.me/bitcoinphantombtcx)
* [Discord](https://discord.gg/bitcoinphantom)
* [Twitter](https://www.twitter.com/bitcoinphantom)
* [Facebook](https://www.facebook.com/bitcoinphantom/)
* [Reddit](https://www.reddit.com/r/bitcoinphantom/)

## Wallets

Binary (pre-compiled) wallets are available on all platforms at [https://bitcoinphantom.com](https://bitcoinphantom.com/#wallets).

> **Note:** **Important!** Only download pre-compiled wallets from the official bitcoinphantom website or official Github repos.

> **Note:** For a fresh wallet install you can reduce the blockchain syncing time by downloading [a nightly snapshot](https://bitcoinphantom-blockchain.com/down) and following the [setup instructions](https://bitcoinphantom-blockchain.com/howto).

### Windows Wallet Usage

1. Download the pre-compiled software.
2. Install
3. In windows file explorer, open `c:\Users\XXX\AppData\Roaming\bitcoinphantom` (be sure to change XXX to your windows user)
4. Right click and create a new file `bitcoinphantom.txt`
5. Edit the file to have the following contents (be sure to change the password)

    ```
    rpcuser=bitcoinphantomrpcusername
    rpcpassword=85CpSuCNvDcYsdQU8w621mkQqJAimSQwCSJL5dPT9wQX
    rpcport=11207
    port=11211
    daemon=1
    algo=groestl
    ```

6. Save and close the file
7. Rename the file to `bitcoinphantom.conf`
8. Start the bitcoinphantom-qt program.
9. Open up bitcoinphantom-qt console and run `getinfo` (or `getmininginfo`) to verify settings.

> **Note:** You must re-start the wallet after making changes to `bitcoinphantom.conf`.

### OS X Wallet

1. Download the pre-compiled software.
2. Double click the DMG
3. Drag the bitcoinphantom-Qt to your Applications folder
4. Double click the bitcoinphantom-Qt application to open it.
5. Go grab a :coffee: while it syncs with the blockchain

> **Note:** It may look like it is frozen or hung while it is indexing and syncing the blockchain. It's not. It's chugging away, but currently the UI doesn't give you a lot of feedback on status. We're working to fix that. Syncing takes a while to complete (ie. > 10 minutes or more) so just be patient.

> **Note:** If you want to change your configuration the file is located at `~/Library/Application\ Support\bitcoinphantom\bitcoinphantom.conf`. This isn't required by default.

### Unix Wallet

1. Compile using [Unix instructions](doc/build-unix.md).
2. The wallet GUI is in `./bitcoinphantom/src/qt` and the daemon in `./bitcoinphantom/src`.
3. **Optional** - the binaries to your favorite location. for use by all users, run the following commands:

    ```shell
    sudo cp src/bitcoinphantomd /usr/bin/
    sudo cp src/qt/bitcoinphantom-qt /usr/bin/
    ```

4. Run `./bitcoinphantomd` from wherever you put it. The output from this command will tell you that you need to make a `bitcoinphantom.conf` file and will suggest some good starting values.
5.  Open up your new config file that was created in your home directory in your favorite text editor

    ```shell
    nano ~/.bitcoinphantom/bitcoinphantom.conf
    ```

6. Paste the output from the `bitcoinphantomd` command into the bitcoinphantom.conf like this: (It is recommended to change the password to something unique.)

    ```
    rpcuser=bitcoinphantomrpcusername
    rpcpassword=85CpSuCNvDcYsdQU8w621mkQqJAimSQwCSJL5dPT9wQX
    rpcport=11207
    port=11211
    daemon=1
    algo=groestl
    ```

7. Save the file and exit your editor. If using `nano` type `ctrl + x` on your keyboard and the `y` and hitting enter. This should have created a `bitcoinphantom.conf` file with what you just added.

8. Start the bitcoinphantom daemon again

    ```shell
    ./path/to/bitcoinphantomd
    ```

> **Note:** To check the status of how much of the blockchain has been downloaded (aka synced) type `./path/to/bitcoinphantomd getinfo`.

> **Note**: If you see something like 'Killed (program cc1plus)' run ```dmesg``` to see the error(s)/problems(s). This is most likely caused by running out of resources. You may need to add some RAM or add some swap space.

You can also check out this [Linux Wallet Video Tutorial](https://www.youtube.com/watch?v=WYe75b6RWes).

## Building From Source

* [Unix Instructions](doc/build-unix.md)
* [OS X Instructions](doc/build-osx.md)
* [Windows Instructions](doc/build-windows.md)

## Developer Notes

The Easy Method:

> **Note**: Sometimes linux user permissions are not set up properly, and causes failed compiling in linux. Please ensure your user has access or do the install from root if these problems arise.

```shell
sudo rm -Rf ~/bitcoinphantom  #(if you already have it)
sudo apt-get -y install git && cd ~ && git clone https://github.com/bitcoinphantom/bitcoinphantom && cd bitcoinphantom && sh go.sh
```

The _slightly_ longer version:

1. Install the dependencies. **Note**: If you are on debian, you will also need to `apt-get install libcanberra-gtk-module`.

    ```shell
    sudo add-apt-repository ppa:bitcoin/bitcoin
    sudo apt-get update
    sudo apt-get install \
        libdb4.8-dev libdb4.8++-dev build-essential \
        libtool autotools-dev automake pkg-config libssl-dev libevent-dev \
        bsdmainutils git libboost-all-dev libminiupnpc-dev libqt5gui5 \
        libqt5core5a libqt5dbus5 libevent-dev qttools5-dev \
        qttools5-dev-tools libprotobuf-dev protobuf-compiler libqrencode-dev \
        libseccomp-dev libcap-dev
    ```

2. Clone the git repository and compile the daemon and gui wallet:

    ```shell
    git clone https://github.com/bitcoinphantom/bitcoinphantom && cd bitcoinphantom && ./autogen.sh && ./configure && make
    ```

> **Note**: If you get a "memory exhausted" error, make a swap file. (https://www.digitalocean.com/community/tutorials/how-to-add-swap-space-on-ubuntu-16-04)


### Mac OS X Wallet

> **Note:** This has only been confirmed to work on OS X Sierra (10.12) and OS X High Sierra (10.13) with XCode 9.2 and `Apple LLVM version 9.0.0 (clang-900.0.39.2)`.

1. Ensure you have mysql and boost installed.

    ```shell
    brew install mysql boost
    ```

2. Ensure you have python 2.7 installed and in your path (OS X comes with this by default)

    ```shell
    python --version
    ```

3. Export the required environment variables

    ```shell
    export bitcoinphantom_PLATFORM='mac'
    export CXX=clang++
    export CC=clang
    ```

4. Run your build commands

    ```shell
    ./building/common.sh
    ./building/mac/requirements.sh
    ./building/mac/build.sh
    ```

5. Grab a :coffee: and wait it out

6. Create the `.dmg` file

    ```shell
    ./building/mac/dist.sh
    ```

### Windows Wallet

TODO. Take a look at [building/windows](./building/windows).

## Docker Images

Check out the [`contrib/readme`](https://github.com/bitcoinphantom/bitcoinphantom/tree/master/contrib/docker) for more information.

## Mining

### Solo mining

Instead of joining a mining pool you can use the wallet to mine all by yourself. You need to specify the algorithm (see below) and set the "gen" flag. For instance, in the configuration specify `gen=1`.

### Using different algorithms

To use a specific mining algorithm use the `algo` switch in your configuration file (`.conf` file) or from the command line (like this `--algo=x17`). Here are the possible values:

```
algo=x17
algo=scrypt
algo=groestl
algo=lyra
algo=blake
```

## TestNet

Here is a list of active testnet nodes:
ddvnucmtvyiemiuk.onion (sunerok)

## Donations

We believe in keeping bitcoinphantom free and open. Any donations to help fuel the development effort are greatly appreciated! :smile:

* Address for donations in bitcoinphantom (btcx): `DDd1pVWr8PPAw1z7DRwoUW6maWh5SsnCcp`
* Address for donations in Bitcoin (BTC): `142r3vCAH3AzABiQjFPmcrSCp6TDzEDuB1`

## Special Shout Outs

Special thanks to the following people that have helped make bitcoinphantom possible. :raised_hands:

Sunerok, CryptoRekt, MKinney, BearSylla, Hypermist, Pallas1, FuzzBawls, BuZz, glodfinch, InfernoMan, AhmedBodi, BitSpill, MentalCollatz, ekryski and the **entire** #bitcoinphantom community!




# Bug Reporting

If you think you've found a bug or a problem with bitcoinphantom, please let us know! First, search our issue tracker to see if someone has already reported the problem. If they haven't, open a new issue, and fill out the template with as much information as possible. The more you can tell us about the problem and how it occurred, the more likely we are to fix it.

## _Please do not report security vulnerabilities publicly._


## How to report a bug

### Code issues

Since we are a 100% open-source project we strongly prefer if you create a pull-request on Github in the proper repository with the necessary fix.

Alternatively, if you would like to make a suggestion regarding a potential fix please send an email to contact@bitcoinphantom.com


### Security-related issues

Contact the developers privately by sending an e-mail to contact@bitcoinphantom.com with the details of the issue. Do not post the issue on github or anywhere else until the issue has been resolved.


doc/release-notes.md
(note: this is a temporary file, to be added-to by anybody, and moved to
release-notes at release time)

bitcoinphantom Core version *version* is now available from:

  <https://github.com/bitcoinphantom/bitcoinphantom/releases>

This is a new major version release, including new features, various bugfixes
and performance improvements, as well as updated translations.

Please report bugs using the issue tracker at GitHub:

  <https://github.com/bitcoinphantom/bitcoinphantom/issues>

How to Upgrade
==============

If you are running an older version, shut it down. Wait until it has completely
shut down (which might take a few minutes for older versions), then run the
installer (on Windows) or just copy over `/Applications/bitcoinphantom-Qt` (on Mac)
or `bitcoinphantomd`/`bitcoinphantom-qt` (on Linux).

The first time you run version 0.15.0, your chainstate database will be converted to a
new format, which will take anywhere from a few minutes to half an hour,
depending on the speed of your machine.

Note that the block database format also changed in version 0.8.0 and there is no
automatic upgrade code from before version 0.8 to version 0.15.0. Upgrading
directly from 0.7.x and earlier without redownloading the blockchain is not supported.
However, as usual, old wallet versions are still supported.

Downgrading warning
-------------------

The chainstate database for this release is not compatible with previous
releases, so if you run 0.15 and then decide to switch back to any
older version, you will need to run the old release with the `-reindex-chainstate`
option to rebuild the chainstate data structures in the old format.

If your node has pruning enabled, this will entail re-downloading and
processing the entire blockchain.

Compatibility
==============

bitcoinphantom Core is extensively tested on multiple operating systems using
the Linux kernel, macOS 10.8+, and Windows 7 and newer (Windows XP is not supported).

bitcoinphantom Core should also work on most other Unix-like systems but is not
frequently tested on them.

Notable changes
===============

GUI changes
-----------

- Block storage can be limited under Preferences, in the Main tab. Undoing this setting requires downloading the full blockchain again. This mode is incompatible with -txindex and -rescan.

RPC changes
------------

### Low-level changes

- The `createrawtransaction` RPC will now accept an array or dictionary (kept for compatibility) for the `outputs` parameter. This means the order of transaction outputs can be specified by the client.
- The `fundrawtransaction` RPC will reject the previously deprecated `reserveChangeKey` option.
- `sendmany` now shuffles outputs to improve privacy, so any previously expected behavior with regards to output ordering can no longer be relied upon.
- The new RPC `testmempoolaccept` can be used to test acceptance of a transaction to the mempool without adding it.
- JSON transaction decomposition now includes a `weight` field which provides
  the transaction's exact weight. This is included in REST /rest/tx/ and
  /rest/block/ endpoints when in json mode. This is also included in `getblock`
  (with verbosity=2), `listsinceblock`, `listtransactions`, and
  `getrawtransaction` RPC commands.
- New `fees` field introduced in `getrawmempool`, `getmempoolancestors`, `getmempooldescendants` and
   `getmempoolentry` when verbosity is set to `true` with sub-fields `ancestor`, `base`, `modified`
   and `descendant` denominated in BTC. This new field deprecates previous fee fields, such as
   `fee`, `modifiedfee`, `ancestorfee` and `descendantfee`.

External wallet files
---------------------

The `-wallet=<path>` option now accepts full paths instead of requiring wallets
to be located in the -walletdir directory.

Newly created wallet format
---------------------------

If `-wallet=<path>` is specified with a path that does not exist, it will now
create a wallet directory at the specified location (containing a wallet.dat
data file, a db.log file, and database/log.?????????? files) instead of just
creating a data file at the path and storing log files in the parent
directory. This should make backing up wallets more straightforward than
before because the specified wallet path can just be directly archived without
having to look in the parent directory for transaction log files.

For backwards compatibility, wallet paths that are names of existing data files
in the `-walletdir` directory will continue to be accepted and interpreted the
same as before.

Low-level RPC changes
---------------------

- When bitcoinphantom is not started with any `-wallet=<path>` options, the name of
  the default wallet returned by `getwalletinfo` and `listwallets` RPCs is
  now the empty string `""` instead of `"wallet.dat"`. If bitcoinphantom is started
  with any `-wallet=<path>` options, there is no change in behavior, and the
  name of any wallet is just its `<path>` string.
- Passing an empty string (`""`) as the `address_type` parameter to
  `getnewaddress`, `getrawchangeaddress`, `addmultisigaddress`,
  `fundrawtransaction` RPCs is now an error. Previously, this would fall back
  to using the default address type. It is still possible to pass null or leave
  the parameter unset to use the default address type.

- Bare multisig outputs to our keys are no longer automatically treated as
  incoming payments. As this feature was only available for multisig outputs for
  which you had all private keys in your wallet, there was generally no use for
  them compared to single-key schemes. Furthermore, no address format for such
  outputs is defined, and wallet software can't easily send to it. These outputs
  will no longer show up in `listtransactions`, `listunspent`, or contribute to
  your balance, unless they are explicitly watched (using `importaddress` or
  `importmulti` with hex script argument). `signrawtransaction*` also still
  works for them.

### Logging


go.sh
#!/bin/bash

#// full deployement :   wget -O - https://raw.githubusercontent.com/badbrainIRC/bitcoinphantom/master/go.sh | bash
sudo rm -Rf ~/bitcoinphantom
# generating entropy make it harder to guess the randomness!.
echo "Initializing random number generator..."
random_seed=/var/run/random-seed
# Carry a random seed from start-up to start-up
# Load and then save the whole entropy pool
if [ -f $random_seed ]
then
    sudo cat $random_seed >/dev/urandom
else
    sudo touch $random_seed
fi
sudo chmod 600 $random_seed
poolfile=/proc/sys/kernel/random/poolsize
[ -r $poolfile ] && bytes=`sudo cat $poolfile` || bytes=512
sudo dd if=/dev/urandom of=$random_seed count=1 bs=$bytes

#Also, add the following lines in an appropriate script which is run during the$

# Carry a random seed from shut-down to start-up
# Save the whole entropy pool
echo "Saving random seed..."
random_seed=/var/run/random-seed
sudo touch $random_seed
sudo chmod 600 $random_seed
poolfile=/proc/sys/kernel/random/poolsize
[ -r $poolfile ] && bytes=`sudo cat $poolfile` || bytes=512
sudo dd if=/dev/urandom of=$random_seed count=1 bs=$bytes

# Create a swap file

cd ~
if [ -e /swapfile1 ]
then
echo "Swapfile already present"
else
sudo dd if=/dev/zero of=/swapfile1 bs=1024 count=1024288
sudo mkswap /swapfile1
sudo chown root:root /swapfile1
sudo chmod 0600 /swapfile1
sudo swapon /swapfile1
fi

# Install dependency

sudo apt-get -y install software-properties-common

sudo add-apt-repository -y ppa:bitcoin/bitcoin

sudo apt-get update

sudo apt-get -y install libcanberra-gtk-module

# Dont need to check if bd is already installed, will override or pass by
#results=$(find /usr/ -name libdb_cxx.so)
#if [ -z $results ]; then
sudo apt-get -y install libdb4.8-dev libdb4.8++-dev
#else
#grep DB_VERSION_STRING $(find /usr/ -name db.h)
#echo "BerkeleyDb will not be installed its already there...."
#fi

sudo apt-get -y install libtool autotools-dev automake pkg-config libssl-dev libevent-dev

sudo apt-get -y install bsdmainutils git libboost-all-dev libminiupnpc-dev libqt5gui5

sudo apt-get -y install libqt5core5a libqt5dbus5 libevent-dev qttools5-dev

sudo apt-get -y install qttools5-dev-tools libprotobuf-dev protobuf-compiler libqrencode-dev

sudo apt-get -y install zlib1g-dev libseccomp-dev libcap-dev libncap-dev

sudo apt-get -y install libunivalue-dev libzmq3-dev

sudo apt-get -y install g++ build-essential

# Keep current version of libboost if already present
results=$(find /usr/ -name libboost_chrono.so)
if [ -z $results ]
then
sudo apt-get -y install libboost-all-dev
else
red=`tput setaf 1`
green=`tput setaf 2`
reset=`tput sgr0`
echo "${green}Libboost will not be installed its already there....${reset}"
grep --include=*.hpp -r '/usr/' -e "define BOOST_LIB_VERSION"
fi

sudo apt-get -y install --no-install-recommends gnome-panel

sudo apt-get -y install lynx

sudo apt-get -y install unzip

sudo apt-get -y install sed

cd ~

#// Compile Berkeley if 4.8 is not there
if [ -e /usr/lib/libdb_cxx-4.8.so ]
then
red=`tput setaf 1`
green=`tput setaf 2`
reset=`tput sgr0`
echo "${green}BerkeleyDb already present...$(grep --include *.h -r '/usr/' -e 'DB_VERSION_STRING')${reset}" 
else
wget http://download.oracle.com/berkeley-db/db-4.8.30.NC.tar.gz 
tar -xzvf db-4.8.30.NC.tar.gz
result2=$(cat /etc/issue | grep -Po '19.04')
if [ $result2 = "19.04" ]
then
sed -i 's/__atomic_compare_exchange/__db_atomic_compare_exchange/g' ~/db-4.8.30.NC/dbinc/atomic.h
fi
result3=$(cat /etc/issue | grep -Po '4.6')
if [ $result3 = "4.6" ]
then
sed -i 's/__atomic_compare_exchange/__db_atomic_compare_exchange/g' ~/db-4.8.30.NC/dbinc/atomic.h
fi
result4=$(cat /etc/issue | grep -Po 'Ermine')
if [ $result4 = "Ermine" ]
then
sed -i 's/__atomic_compare_exchange/__db_atomic_compare_exchange/g' ~/db-4.8.30.NC/dbinc/atomic.h
fi

rm db-4.8.30.NC.tar.gz
cd db-4.8.30.NC/build_unix 
../dist/configure --enable-cxx
make 
sudo make install 
sudo ln -s /usr/local/BerkeleyDB.4.8/lib/libdb-4.8.so /usr/lib/libdb-4.8.so
sudo ln -s /usr/local/BerkeleyDB.4.8/lib/libdb_cxx-4.8.so /usr/lib/libdb_cxx-4.8.so
cd ~
sudo rm -Rf db-4.8.30.NC
#sudo ldconfig
fi

#// Check if libboost is present

results=$(find /usr/ -name libboost_chrono.so)
if [ -z $results ]
then
sudo rm download
     wget https://sourceforge.net/projects/boost/files/boost/1.67.0/boost_1_67_0.zip/download 
     unzip -o download
     cd boost_1_67_0
	sh bootstrap.sh
	sudo ./b2 install
	cd ~
	sudo rm download 
	sudo rm -Rf boost_1_67_0
	#sudo ln -s $(dirname "$(find /usr/ -name libboost_chrono.so)")/lib*.so /usr/lib
	sudo ldconfig
        #sudo rm /usr/lib/libboost_chrono.so
else
     echo "Libboost found..." 
     grep --include=*.hpp -r '/usr/' -e "define BOOST_LIB_VERSION"
fi

#// Clone files from repo, Permissions and make

git clone --recurse-submodules https://github.com/bitcoinphantom/bitcoinphantom
cd ~
cd bitcoinphantom
./autogen.sh
chmod 777 ~/bitcoinphantom/share/genbuild.sh
chmod 777 ~/bitcoinphantom/src/leveldb/build_detect_platform

grep --include=*.hpp -r '/usr/' -e "define BOOST_LIB_VERSION"

sudo rm wrd01.txt
sudo rm wrd00.txt
sudo rm words
find /usr/ -name libboost_chrono.so > words
split -dl 1 --additional-suffix=.txt words wrd

if [ -e wrd01.txt ]
then
echo 0. $(cat wrd00.txt)
echo 1. $(cat wrd01.txt)
echo 2. $(cat wrd02.txt)
echo 3. $(cat wrd03.txt)
echo -n "Choose libboost library to use(0-3)?"
read answer
else
echo "There is only 1 libboost library present. We choose for you 0"
answer=0
fi

echo "You have choosen $answer"

if [ $(dirname "$(cat wrd00.txt)") = "/usr/lib/arm-linux-gnueabihf" ]
then
red=`tput setaf 1`
green=`tput setaf 2`
reset=`tput sgr0`
echo "${green}ARM cpu detected --disable-sse2${reset}"
txt=$(echo "--disable-sse2")
sed -i 's/#include <emmintrin.h>//g' ~/bitcoinphantom/src/crypto/pow/scrypt-sse2.cpp
else
txt=$(echo "")
fi

if [ -d /usr/local/BerkeleyDB.4.8/include ]
then
cd bitcoinphantom
./configure CXXFLAGS="--param ggc-min-expand=1 --param ggc-min-heapsize=32768" CPPFLAGS="-I/usr/local/BerkeleyDB.4.8/include -O2" LDFLAGS="-L/usr/local/BerkeleyDB.4.8/lib" --with-gui=qt5 --with-boost-libdir=$(dirname "$(cat wrd0$answer.txt)") --disable-bench --disable-tests --disable-gui-tests --without-miniupnpc $txt
echo "Using Berkeley Generic..."
else
cd bitcoinphantom
./configure CXXFLAGS="--param ggc-min-expand=1 --param ggc-min-heapsize=32768" CPPFLAGS="-O2" --with-gui=qt5 --with-boost-libdir=$(dirname "$(cat wrd0$answer.txt)") --disable-bench --disable-tests --disable-gui-tests --without-miniupnpc $txt

echo "Using default system Berkeley..."
fi

#make -j$(nproc) USE_UPNP=-
make USE_UPNP=-

if [ -e ~/bitcoinphantom/src/qt/bitcoinphantom-qt ]
then
sudo strip ~/bitcoinphantom/src/bitcoinphantomd
sudo strip ~/bitcoinphantom/src/qt/bitcoinphantom-qt
sudo make install
else
echo "Compile fail not bitcoinphantom-qt present"
fi

cd ~

#// Create the config file with random user and password

mkdir -p ~/.bitcoinphantom
if [ -e ~/.bitcoinphantom/bitcoinphantom.conf ]
then
    cp -a ~/.bitcoinphantom/bitcoinphantom.conf ~/.bitcoinphantom/bitcoinphantom.bak
fi
rm ~/.bitcoinphantom/bitcoinphantom.conf
echo "rpcuser="$(head /dev/urandom | tr -dc A-Za-z0-9 | head -c 26) >> ~/.bitcoinphantom/bitcoinphantom.conf
echo "rpcpassword="$(head /dev/urandom | tr -dc A-Za-z0-9 | head -c 26) >> ~/.bitcoinphantom/bitcoinphantom.conf
echo "rpcport=11207" >> ~/.bitcoinphantom/bitcoinphantom.conf
echo "port=11211" >> ~/.bitcoinphantom/bitcoinphantom.conf
echo "daemon=1" >> ~/.bitcoinphantom/bitcoinphantom.conf
echo "listen=1" >> ~/.bitcoinphantom/bitcoinphantom.conf
echo "server=1" >> ~/.bitcoinphantom/bitcoinphantom.conf
echo "deprecatedrpc=accounts" >> ~/.bitcoinphantom/bitcoinphantom.conf

echo "usehd=1" >> ~/.bitcoinphantom/bitcoinphantom.conf


#// Extract http link, download blockchain and install it.

sudo chmod 777 /etc/lynx/lynx.cfg
sudo echo "FORCE_SSL_PROMPT:YES" >> /etc/lynx/lynx.cfg

# Create Icon on Desktop and in menu
mkdir -p ~/Desktop/
sudo cp ~/bitcoinphantom/src/qt/res/icons/bitcoinphantom.png /usr/share/icons/
echo "#!/usr/bin/env xdg-open" >> ~/Desktop/bitcoinphantom.desktop
echo "[Desktop Entry]" >> ~/Desktop/bitcoinphantom.desktop
echo "Version=1.0" >> ~/Desktop/bitcoinphantom.desktop
echo "Type=Application" >> ~/Desktop/bitcoinphantom.desktop
echo "Terminal=false" >> ~/Desktop/bitcoinphantom.desktop
echo "Icon[en]=/usr/share/icons/bitcoinphantom.png" >> ~/Desktop/bitcoinphantom.desktop
echo "Name[en]=bitcoinphantom" >> ~/Desktop/bitcoinphantom.desktop
echo "Exec=bitcoinphantom-qt" >> ~/Desktop/bitcoinphantom.desktop
echo "Name=bitcoinphantom" >> ~/Desktop/bitcoinphantom.desktop
echo "Icon=/usr/share/icons/bitcoinphantom.png" >> ~/Desktop/bitcoinphantom.desktop
echo "Categories=Network;Internet;" >> ~/Desktop/bitcoinphantom.desktop
sudo chmod +x ~/Desktop/bitcoinphantom.desktop
sudo cp ~/Desktop/bitcoinphantom.desktop /usr/share/applications/bitcoinphantom.desktop
sudo chmod +x /usr/share/applications/bitcoinphantom.desktop

# Erase all bitcoinphantom compilation directory , cleaning

cd ~
#sudo rm -Rf ~/bitcoinphantom

# Blockchain

echo -n "Success....Blockchain is now downloading press Ctrl-C to cancel but it will take longer to sync from 0. And you will have to start bitcoinphantom manual"
sudo rm QT-Wallet*.zip
echo "wget --no-check-certificate " $(lynx --dump --listonly https://bitcoinphantom-blockchain.com/down/ | grep -o "https://bitcoinphantom-blockchain.com/blockchain5.*zip" | head -1 ) > link.sh
sh link.sh
unzip -o QT-Wallet*.zip -d ~/.bitcoinphantom
sudo rm QT-Wallet*.zip
#// Start bitcoinphantom
bitcoinphantom-qt
